baseapp/abci.go

```diff
@@ -280,7 +297,21 @@ func (app *BaseApp) ProcessProposal(req abci.RequestProcessProposal) abci.Respon
 		WithProposer(req.ProposerAddress).
 		WithConsensusParams(app.GetConsensusParams(app.processProposalState.ctx))
 
-	return app.processProposal(ctx, req)
+	defer func() {
+		if err := recover(); err != nil {
+			app.logger.Error(
+				"panic recovered in ProcessProposal",
+				"height", req.Height,
+				"time", req.Time,
+				"hash", fmt.Sprintf("%X", req.Hash),
+				"panic", err,
+			)
+			resp = abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}
```

Comment:
I mean it is still better than stopping the node, but it is still stuck then?

Reply:
Since `ProcessProposal` is defined by the application and could very well define behavior that depends on the proposal's contents, i.e. txs, then the behavior cannot be guaranteed to be deterministic.

That being said, if a panic is triggered, the proposal is rejected. As such, Tendermint will proceed to the next validator to propose a new proposal in PrepareProposal and the round starts again. At some point a valid proposal will need to be accepted. If no such proposal occurs, then yes, the chain will face a liveness halt, but this will be very clear and investigation will need to occur.

Reply:
Just a quick followup in case folks come back to this comment in the future:

- `PrepareProposal` MAY be non-deterministic
- `ProcessProposal` MAY NOT be non-deterministic

This means if `ProcessProposal` panics and we reject, the validator process `p` will prevote nil and the round will proceed again until a valid proposal is proposed.

Reply:
we should document this here: https://docs.cosmos.network/main/building-apps/app-mempool

Reply:
https://github.com/cosmos/cosmos-sdk/pull/14393